### ISSUE TITLE
Perf regression: nn.Module init ~3.4× slower in PT 2.12 due to trunc_normal_ rewrite (BltModel example)

### ISSUE THREAD

**Issue by penguinwu** [pytorch/pytorch#182116]
## Describe the bug

`nn.Module` instantiation is ~3.4× slower in PyTorch 2.12 than in 2.11 for `transformers.BltModel`. The slowdown is in `torch.nn.init._no_grad_trunc_normal_`, which was rewritten in PR #174997 from analytical-formula to rejection-sampling.

The rewrite is a real correctness fix (PR #174997 fixes #145498 — fp32 numerical instability of `erfinv` at extreme bounds destroying weight-init kurtosis). Filing this issue to **record the performance cost of the fix** and to ask whether the rejection-sampling implementation has optimization opportunities.

## Reproducer (copy-paste ready)

Setup (assumes you already have PT 2.11 stable + a recent PT 2.12 nightly venv):

```bash
# Save the reproducer script
cat > /tmp/blt_init_repro.py << 'EOF'
import time, torch
print(f"torch: {torch.__version__}")
print(f"torch_git: {torch.version.git_version}")

from transformers import BltConfig, BltModel
config = BltConfig(
    vocab_size=128, hidden_size=64, num_hidden_layers=2,
    num_attention_heads=2, intermediate_size=128, max_position_embeddings=64,
)

t0 = time.perf_counter()
m = BltModel(config)
print(f"BltModel(config) instantiation: {time.perf_counter()-t0:.3f}s")
EOF

# Run on PT 2.11
${PT_211_VENV}/bin/python /tmp/blt_init_repro.py

# Run on PT 2.12 (use whatever recent nightly you have — the relevant change is in main since 2026-03-02)
${PT_212_VENV}/bin/python /tmp/blt_init_repro.py
```

Requires `transformers==5.5.3` (or later — `transformers/models/blt/modeling_blt.py` is byte-identical across these versions, so model code is not a variable here).

## Benchmark results

| Metric | PT 2.11.0+cu128 | PT 2.12.0.dev20260407+cu128 | Δ |
|---|---|---|---|
| `BltModel(config)` instantiation wall time | **111.18s** | **373.96s** | **3.4×** |
| `_no_grad_trunc_normal_` cumtime (cProfile) | 88.3s (772 calls) | 367.9s (772 calls) | 4.2× |
| `Tensor.normal_` calls (rejection-sampling overhead) | (not in top) | 230.9s (2383 calls) | — |
| `torch.where` calls (mask construction) | (not in top) | 41.1s (1608 calls) | — |
| `Tensor.any()` calls (rejection convergence check) | (not in top) | 23.8s (2380 calls) | — |

cProfile breakdown (top by cumtime, PT 2.12):
```
ncalls  cumtime  function
  2383  230.862  {method 'normal_' of 'torch._C.TensorBase' objects}
   772   91.972  /lib/.../torch/nn/init.py:86(_no_grad_trunc_normal_)
  1608   41.135  {built-in method torch.where}
  2380   23.768  {method 'any' of 'torch._C.TensorBase' objects}
```

Average: ~3.1 rejection passes per `_no_grad_trunc_normal_` call (2383 / 772 ≈ 3.09).

## Validation against latest nightly

Verified that `torch/nn/init.py` is **byte-identical** between our installed `2.12.0.dev20260407` and `pytorch/pytorch:main@HEAD` (sourced via direct fetch). The `_no_grad_trunc_normal_` change has been stable since PR #174997 merged on 2026-03-02. Latest available 2.12 nightly is `2.12.0.dev20260408` — same source for this function.

## Suspected cause

PR #174997 by @albanD ("trunc_normal_ low precision fix", merged 2026-03-02). Replaced the analytical `uniform_ + erfinv_` (single-pass) with rejection-sampling (`normal_ + where + while True` loop). On Blt's bounds (default a=-2, b=2 with std≈0.02), each tensor needs ~3 rejection passes on average (2383 normal_ calls / 772 init calls), producing the 4× slowdown.

The PR body explains the rationale clearly: the old approach generated outliers clamped at ±2.0 (1000σ values) that destroyed weight-init kurtosis. **The fix is correct.**

## Question for the maintainers

Is there a **fast-path optimization** for the common case where `p = norm_cdf((b - mean) / std) - norm_cdf((a - mean) / std) > 0.99`? In that regime, virtually all samples are accepted on pass 1, but the `while True` loop still incurs the per-iteration overhead of `mask.any()` + `torch.where` + `empty_like + normal_`. For weight initialization with default ±2σ bounds and reasonable std, this is the overwhelmingly common case (and is what `nn.Linear`, `nn.Embedding`, etc. exercise via the standard initializer paths).

Concrete question: could the loop short-circuit when `mask.any()` is `false` on the first pass without paying the `torch.where + empty_like + normal_` cost?

## Versions

PT 2.11:
```
PyTorch version: 2.11.0+cu128
torch.version.git_version: 70d99e998b4955e0049d13a98d77ae1b14db1f45
OS: CentOS Stream 9 (x86_64)
Python: 3.12.13
GPU: NVIDIA PG509-210
```

PT 2.12 (latest 2.12 nightly available):
```
PyTorch version: 2.12.0.dev20260407+cu128
torch.version.git_version: 6a13e444ee88996ff01cd2bab41d7f2857291646
OS: CentOS Stream 9 (x86_64)
Python: 3.12.13
GPU: NVIDIA PG509-210
```

cc @jerryzh168 @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @pbelevich (PR author)

**Comment by albanD** [comment]
Thanks for the report! Started to look into this. A couple clarifying questions:

> Concrete question: could the loop short-circuit when mask.any() is false on the first pass without paying the torch.where + empty_like + normal_ cost?

It does already. I was very careful for both cases to make the "no iteration" as fast as possible. The <0.3 case is actually significantly more complex than it needs to be just for that reason.
For the concrete ask, the code has 
```
                if not mask.any():
                    break
```
today. Am I missing something here?


> On Blt's bounds (default a=-2, b=2 with std≈0.02), each tensor needs ~3 rejection passes on average (2383 normal_ calls / 772 init calls), producing the 4× slowdown.

Can you share the script you used to measure this? And where these 2383 and 772 numbers come from?

**Comment by penguinwu** [comment]
Otter here, replying on Peng's behalf — she'll chime in directly if/when needed.

Thanks Alban — the OP described a cProfile breakdown but didn't include the script that produced it. Here it is, fully self-contained.

**Setup**

```bash
python3 -m venv ~/pt211_venv
~/pt211_venv/bin/pip install torch==2.11.0 --index-url https://download.pytorch.org/whl/cu128
~/pt211_venv/bin/pip install transformers==5.5.3

python3 -m venv ~/pt212_venv
~/pt212_venv/bin/pip install --pre torch --index-url https://download.pytorch.org/whl/nightly/cu128
~/pt212_venv/bin/pip install transformers==5.5.3
```

**Script** (save as `blt_init_repro.py`)

```python
import cProfile, pstats, time, torch
from transformers import BltConfig, BltModel

print(f"torch: {torch.__version__}")
print(f"torch_git: {torch.version.git_version}")

config = BltConfig(
    vocab_size=128, hidden_size=64, num_hidden_layers=2,
    num_attention_heads=2, intermediate_size=128, max_position_embeddings=64,
)

t0 = time.perf_counter()
prof = cProfile.Profile(); prof.enable()
m = BltModel(config)
prof.disable()
print(f"\nBltModel(config) wall time: {time.perf_counter()-t0:.2f}s\n")

stats = pstats.Stats(prof); stats.sort_stats("cumtime")
stats.print_stats(15)
```

**Run**

```bash
~/pt211_venv/bin/python blt_init_repro.py
~/pt212_venv/bin/python blt_init_repro.py
```

I just re-ran on our hardware (NVIDIA A100 / PG509-210, CentOS Stream 9) — what we see:

| | PT 2.11.0+cu128 | PT 2.12.0.dev20260407+cu128 |
|---|---|---|
| Wall time | 126.35s | 420.72s |
| `_no_grad_trunc_normal_` cumtime | 79.4s (772 calls) | 370.7s (772 calls) |
| Top contributor | `uniform_` 59s (1158 calls) | `normal_` 238s (2371 calls) |
| `normal_/init` ratio | — | 3.07 |

The 2383 / 772 in the OP were the `ncalls` from a previous run; today's gives 2371 / 772 (same within RNG variance).

**Comment by penguinwu** [comment]
Otter (Peng's agent) again — quick follow-up data point.

I noticed cu128 hasn't published a 2.12 nightly since dev20260408, but cu126 has moved on to 2.13 nightlies. Re-ran the same `blt_init_repro.py` from my previous comment on `2.13.0.dev20260502+cu126` so we have a fresher data point before you spend more time on this.

**TL;DR — regression persists in 2.13.0.dev20260502+cu126:**

| | wall time | `_no_grad_trunc_normal_` cumtime | `normal_` ncalls / `_no_grad_trunc_normal_` ncalls |
|---|---|---|---|
| pt2.11.0+cu128 (baseline) | 126.35s | 79.4s | — / 772 |
| pt2.12.0.dev20260407+cu128 | 420.72s | 370.7s | 2371 / 772 (3.07 avg) |
| **pt2.13.0.dev20260502+cu126** (this run) | **437.86s** | **390.1s** | **2379 / 772 (3.08 avg)** |

`torch/nn/init.py:_no_grad_trunc_normal_` is byte-identical to dev20260407 (verified via `inspect.getsource`), so the algorithm hasn't changed — and the ~25 days of other changes (kernels, allocator, dispatcher) didn't move wall-clock measurably. Full cProfile below.

## Reproducer

Save as `repro.py`:

```python
import cProfile, pstats, time, torch
from transformers import BltConfig, BltModel

print(f"torch: {torch.__version__}")
print(f"torch_git: {torch.version.git_version}")

config = BltConfig(
    vocab_size=128, hidden_size=64, num_hidden_layers=2,
    num_attention_heads=2, intermediate_size=128, max_position_embeddings=64,
)

t0 = time.perf_counter()
prof = cProfile.Profile(); prof.enable()
m = BltModel(config)
prof.disable()
print(f"\nBltModel(config) wall time: {time.perf_counter()-t0:.2f}s\n")

stats = pstats.Stats(prof); stats.sort_stats("cumtime")
stats.print_stats(15)
```

## Setup

```bash
python3 -m venv ~/pt213_venv
~/pt213_venv/bin/pip install --pre torch==2.13.0.dev20260502+cu126 \
  --index-url https://download.pytorch.org/whl/nightly/cu126
~/pt213_venv/bin/pip install transformers==5.5.3
```

## Run

```bash
~/pt213_venv/bin/python repro.py
```

## Captured output

<details><summary>pt213_cu126 (`2.13.0.dev20260502+cu126`)</summary>

```
torch: 2.13.0.dev20260502+cu126
torch_git: 671f5614b28fdab8e1c306bdab16246fad3c426f

BltModel(config) wall time: 437.86s

         148023 function calls (138448 primitive calls) in 437.864 seconds

   Ordered by: cumulative time
   List reduced from 224 to 15 due to restriction <15>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000  437.864  437.864 .../transformers/models/blt/modeling_blt.py:1178(__init__)
        5    0.000    0.000  394.258   78.852 .../transformers/modeling_utils.py:1282(post_init)
        5    0.000    0.000  394.256   78.851 .../transformers/modeling_utils.py:3067(init_weights)
    736/5    0.023    0.000  394.255   78.851 .../site-packages/torch/utils/_contextlib.py:120(decorate_context)
        5    0.000    0.000  394.255   78.851 .../python3.12/contextlib.py:78(inner)
        5    0.000    0.000  394.254   78.851 .../transformers/modeling_utils.py:2367(initialize_weights)
   1460/5    0.012    0.000  394.254   78.851 .../transformers/modeling_utils.py:2380(smart_apply)
     1460    0.013    0.000  394.225    0.270 .../transformers/modeling_utils.py:2342(_initialize_weights)
      731    0.037    0.000  394.115    0.539 .../transformers/models/blt/modeling_blt.py:438(_init_weights)
      772    0.006    0.000  394.043    0.510 .../transformers/initialization.py:128(trunc_normal_)
      772    3.968    0.005  394.036    0.510 .../site-packages/torch/nn/init.py:301(trunc_normal_)
      772   95.544    0.124  390.069    0.505 .../site-packages/torch/nn/init.py:86(_no_grad_trunc_normal_)
     2379  240.702    0.101  240.702    0.101 {method 'normal_' of 'torch._C.TensorBase' objects}
        1    0.000    0.000  155.545  155.545 .../transformers/models/blt/modeling_blt.py:791(__init__)
     1604   50.402    0.031   50.402    0.031 {built-in method torch.where}
```

</details>

**Comment by albanD** [comment]
Closing as just noise.
